### PR TITLE
bpo-35059: Convert PyObject_INIT() to function

### PR DESCRIPTION
**Pull request by vstinner**
* Convert PyObject_INIT() and PyObject_INIT_VAR() macros to static
  inline functions.
* Fix usage of these functions.

https://bugs.python.org/issue35059

### CODE REVIEW

**Comment by vstinner** [comment]
Strange, AppVeyor seems stuck: " continuous-integration/appveyor/pr Expected — Waiting for status to be reported ". I will close/reopen the issue to schedule a new job.

**Comment by vstinner** [comment]
My PR #10079 adds a new Py_STATIC_INLINE() macro. I will wait until the other PR is merged to be able to use Py_STATIC_INLINE() in this PR.

**Comment by vstinner** [comment]
Ok, I merged PR #10093 which adds Py_STATIC_INLINE(). I rebased this PR on top of it and I modified this PR to use Py_STATIC_INLINE().

I also chose to remove the NEWS entry.

@methane: Would you mind to review it?

**Comment by methane** [comment]
Doesn't this break many third party modules which use `PyObject_INIT()`?

**Comment by methane** [comment]
> Which kind of _failure_ do you expect?

You have added some `(PyObject*)` cast in caller.
I expect same fix is required for third party extensions.

One option to avoid it is using `void*` as first argument and cast it to `PyObject*` in the function.
But it kills one of the merits of inline function over macros.

**Comment by vstinner** [comment]
>  I expect same fix is required for third party extensions.

Right. Do you consider that it's an issue? It's just a compiler warning. IMHO it's a good thing that developers cast explicitly, no? A function (macro currently) which accepts any type is just weird.

If you prefer, I can rename the macro to _PyObject_INIT() and add:

```
#define PyObject_INIT(obj) _PyObject_INIT((PyObjcet *)(obj))
```

So the "function" accepts any type, as your void* suggestion.

@methane: what do you think?

**Comment by methane** [comment]
> Right. Do you consider that it's an issue? It's just a compiler warning.

I expected it cause compile error, not warning.
I prefer current, simple and clean solution.

**Review by methane — APPROVED** [review]

**Comment by vstinner** [comment]
> I expected it cause compile error, not warning.

It only causes the compilation to fail if you use -Werror. Very few people use this option, and if they use it, they are well aware of the effect of the flag :-)

**Comment by vstinner** [comment]
> I prefer current, simple and clean solution.

If too many users complain, we can add the macro doing the cast later. Even after the 3.8.0 release.

**Pull request merged**
